File: 0101-symmetric-tree/0101-symmetric-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        if (root == nullptr) return true;
        else return subsymmetric(root->left, root->right);
    }
    bool subsymmetric(TreeNode* leftsub, TreeNode* rightsub){
        if (leftsub == nullptr and rightsub == nullptr) return true;
        else if (leftsub == nullptr or rightsub == nullptr) return false;
        if (leftsub -> val != rightsub->val) return false;   
        else return subsymmetric(leftsub->left, rightsub->right) and subsymmetric(leftsub->right, rightsub->left);
    }
};

File: 0101-symmetric-tree/0101-symmetric-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return False
        return self.subtree(root.left, root.right)
            
        
    def subtree(self, leftp, rightp):
        if leftp == None and rightp == None: return True
        elif leftp == None or rightp == None: return False
        if leftp.val != rightp.val : return False
        else: 
            return self.subtree(leftp.left, rightp.right) and self.subtree(leftp.right, rightp.left)
